datasets/noaa-cdr/noaa_cdr.py

```diff
@@ -0,0 +1,189 @@
+import logging
+import os.path
+from tempfile import TemporaryDirectory
+from typing import List, Union
+
+import stactools.noaa_cdr.constants
+import stactools.noaa_cdr.ocean_heat_content.stac
+import stactools.noaa_cdr.sea_ice_concentration.stac
+import stactools.noaa_cdr.sea_surface_temperature_optimum_interpolation.stac
+import stactools.noaa_cdr.sea_surface_temperature_whoi.stac
+import stactools.noaa_cdr.stac
+from pystac import Item, MediaType
+
+from pctasks.core.models.task import WaitTaskResult
+from pctasks.core.storage import StorageFactory
+from pctasks.dataset.collection import Collection
+
+logger = logging.getLogger(__name__)
+
+
+def cog_uri(name: str) -> str:
+    return f"blob://noaacdr/cogs/{name}"
+
+
+class OceanHeatContentCollection(Collection):
+    @classmethod
+    def create_item(
+        cls, asset_uri: str, storage_factory: StorageFactory
+    ) -> Union[List[Item], WaitTaskResult]:
+        directory = "/".join(asset_uri.split("/")[:-1])
+        asset_storage = storage_factory.get_storage(directory)
+        files = asset_storage.list_files(extensions=[".nc"])
+        cog_storage = storage_factory.get_storage(cog_uri("ocean-heat-content"))
+        cog_files = cog_storage.list_files(extensions=[".tif"])
+        cog_hrefs = [cog_storage.get_url(file) for file in cog_files]
+        with TemporaryDirectory() as temporary_directory:
+            local_netcdf_paths = []
+            for file in files:
+                path = os.path.join(temporary_directory, os.path.basename(file))
+                asset_storage.download_file(file, path)
+                local_netcdf_paths.append(path)
+            items = stactools.noaa_cdr.ocean_heat_content.stac.create_items(
+                hrefs=local_netcdf_paths,
+                directory=temporary_directory,
+                cog_hrefs=cog_hrefs,
+            )
+            # TODO add a derived-from link
+            for item in items:
+                for key, asset in item.assets.items():
+                    if asset.href.startswith(
+                        temporary_directory
+                    ) and asset.href.endswith(".tif"):
+                        file_name = os.path.basename(asset.href)
+                        cog_url = cog_storage.get_url(file_name)
+                        cog_storage.upload_file(asset.href, file_name)
+                        asset.href = cog_url
+                        item.assets[key] = asset
+        return items
+
+
+class OceanHeatContentNetcdfCollection(Collection):
+    @classmethod
+    def create_item(
+        cls, asset_uri: str, storage_factory: StorageFactory
+    ) -> Union[List[Item], WaitTaskResult]:
+        asset_storage, asset_path = storage_factory.get_storage_for_file(asset_uri)
+        with TemporaryDirectory() as temporary_directory:
+            path = os.path.join(temporary_directory, os.path.basename(asset_path))
+            asset_storage.download_file(asset_path, path)
```

Comment:
Just confirming that downloading the asset here is necessary. I see that we use fsspec to open the asset at https://github.com/stactools-packages/noaa-cdr/blob/56061afd2d90aaaa3a783e2082228dbece266826/src/stactools/noaa_cdr/ocean_heat_content/stac.py#L188-L190.

I haven't looked at `stac.create_item`, but in theory if we're only getting an attribute out of the NetCDF file we should be able to pass h5py a file-like object like you get from `fsspec.open(https://...)`. But I've had some bad luck with that hanging occasionally, so perhaps we're best off downloading the file.

Reply:
Good point, I think I just blindly did the same thing I was doing while cogifying.

> bad luck with that hanging occasionally, so perhaps we're best off downloading the file.

It's a pretty infrequent update, so rare hangs _shouldn't_ be a huge problem I guess? I'll implement w/o download and see if it works on the test instance.

Reply:
Bailing on fixing this due to offline conversation where it turns out the non-download access might do way more http requests 🤷🏼